crates/red_knot_test/src/parser.rs

```diff
@@ -269,57 +252,142 @@ impl<'s> Parser<'s> {
+        let start = self.offset().to_usize();
+
+        while !self.cursor.is_eof() {
+            let c = self.cursor.first();
+            if end_predicate(c) {
+                return Some(&self.source[start..self.offset().to_usize()]);
+            }
+            self.cursor.bump();
+        }
+
+        None
+    }
+
     fn parse_impl(&mut self) -> anyhow::Result<()> {
-        while let Some(position) = memchr2(b'`', b'#', self.cursor.as_bytes()) {
-            self.cursor.skip_bytes(position.saturating_sub(2));
-
-            // Code blocks and headers must start on a new line
-            // and preceded by at least one blank line.
-            if position == 0 && self.cursor.first() == '#'
-                || position == 1 && self.cursor.eat_char('\n')
-                || self.cursor.eat_char('\n') && self.cursor.eat_char('\n')
-            {
-                match self.cursor.first() {
-                    '#' => {
-                        if let Some(find) = HEADER_RE.find(self.cursor.as_str()) {
-                            self.parse_header(find.as_str())?;
-                            self.cursor.skip_bytes(find.len());
-                            continue;
+        while !self.cursor.is_eof() {
+            self.skip_whitespace();
+
+            match self.cursor.first() {
+                '#' => {
+                    self.explicit_path = None;
+                    self.preceding_blank_lines = 0;
+
+                    // Determine header level (number of '#' characters)
+                    let mut header_level = 0;
+                    while self.cursor.eat_char('#') {
+                        header_level += 1;
+                    }
+
+                    // Parse header title
+                    if let Some(title) = self.consume_until(|c| c == '\n') {
+                        let title = title.trim();
+
+                        if !title.is_empty() {
+                            self.process_header(header_level, title)?;
                         }
                     }
-                    '`' => {
-                        if let Some(captures) = CODE_RE.captures(self.cursor.as_str()) {
-                            self.parse_code_block(&captures)?;
-                            self.cursor.skip_bytes(captures.get(0).unwrap().len());
-                            continue;
+                }
+                '`' => {
+                    self.cursor.eat_char('`');
+
+                    match self.cursor.first() {
+                        '`' if self.cursor.second() == '`' => {
+                            // We see the triple-backtick beginning of a code block.
+                            self.cursor.bump();
+                            self.cursor.bump();
+
+                            if self.preceding_blank_lines < 1 && self.explicit_path.is_none() {
+                                bail!("Code blocks must start on a new line and be preceded by at least one blank line.");
```

Comment:
This message could be improved/adjusted according to the logic in the `if` condition, but I didn't want to introduce and behavioral changes here.